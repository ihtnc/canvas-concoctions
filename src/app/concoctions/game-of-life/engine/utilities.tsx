import { type MatrixCoordinate, type MatrixValue, PeekDirection, peek } from "@/utilities/matrix-operations"
import { isCellUndefinedOrTechnicallyAlive } from "./cell-state-operations"
import { type CellValue, CellState } from "./types"
import { type HSL, type BorderRadii } from "@/utilities/drawing-operations"
import { deepCopy } from "@/utilities/misc-operations"
import ENGINE_DATA from './data'
import { getCycleProgress } from "./cycle-operations"

type GenerateCellsFunction = (value: MatrixValue<CellValue>, coordinate: MatrixCoordinate) => MatrixValue<CellValue>
export const generateCells: GenerateCellsFunction = (value, coordinate) => {
  const { row, col } = coordinate
  if (row >= value.length) { return value }
  if (col >= value[row].length) { return value }

  value[row][col] = deepCopy(ENGINE_DATA.AliveCell)
  return value
}

type CellStateMap = { state: CellState, cells: Array<MatrixCoordinate> }
type GetStateMapFunction = (value: MatrixValue<CellValue>) => Array<CellStateMap>
export const getStateMap: GetStateMapFunction = (value): Array<CellStateMap> => {
  const sorted: Array<CellStateMap> = []

  for (let i = 0; i < value.length; i++) {
    for (let j = 0; j < value[i].length; j++) {
      const state = value[i][j].state
      let group = sorted.find(item => item.state === state)
      if (group === undefined) {
        const newLength = sorted.push({ state, cells: [] })
        group = sorted[newLength - 1]
      }
      group.cells.push({ row: i, col: j })
    }
  }

  return sorted
}

type CellDisplayMap = {
  state: CellState,
  color: HSL,
  opacity: number,
  width: number, height:
  number, cells: Array<MatrixCoordinate>
}
type GetDisplayMapFunction = (stateMap: Array<CellStateMap>, cycleIndex: number, width: number, height: number, aliveColor: HSL, dyingColor: HSL, growingColor: HSL) => Array<CellDisplayMap>
export const getDisplayMap: GetDisplayMapFunction = (stateMap, cycleIndex, width: number, height: number, aliveColor, dyingColor, growingColor): Array<CellDisplayMap> => {
  const sorted: Array<CellDisplayMap> = []
  const cycleProgress = getCycleProgress(cycleIndex)

  for (let i = 0; i < stateMap.length; i++) {
    if (stateMap[i].state === CellState.Dead) { continue }

    let color: HSL
    let cellWidth: number
    let cellHeight: number
    let opacity: number
    const state = stateMap[i].state
    switch (state) {
      case CellState.Dying:
        const dyingProgress = 1 - cycleProgress
        color = deepCopy(dyingColor)
        color.s = color.s * dyingProgress
        opacity = dyingProgress
        cellWidth = width * dyingProgress
        cellHeight = height * dyingProgress
        break
      case CellState.Growing:
        const growingProgress = cycleProgress
        color = deepCopy(growingColor)
        color.s = color.s * growingProgress
        opacity = growingProgress
        cellWidth = width * growingProgress
        cellHeight = height * growingProgress
        break
      default:
        color = deepCopy(aliveColor)
        opacity = 1
        cellWidth = width
        cellHeight = width
        break
    }

    const group: CellDisplayMap = {
      state,
      color,
      opacity,
      width: cellWidth,
      height: cellHeight,
      cells: []
    }
    for (let j = 0; j < stateMap[i].cells.length; j++) {
      group.cells.push(stateMap[i].cells[j])
    }

    sorted.push(group)
  }

  return sorted
}

export const getCellShape: (map: MatrixValue<CellValue>, coordinate: MatrixCoordinate, cycleIndex: number) => BorderRadii = (map, coordinate, cycleIndex) => {
  const { row, col } = coordinate
  const shape = deepCopy(ENGINE_DATA.CellShape)

  const cellTopLeft = peek(map, { row, col }, PeekDirection.UpperLeft)
  if (isCellUndefinedOrTechnicallyAlive(cellTopLeft)) {
    shape.tl = 0
  }

  const cellAbove = peek(map, { row, col }, PeekDirection.Up)
  if (isCellUndefinedOrTechnicallyAlive(cellAbove)) {
    shape.tl = 0
    shape.tr = 0
  }

  const cellTopRight = peek(map, { row, col }, PeekDirection.UpperRight)
  if (isCellUndefinedOrTechnicallyAlive(cellTopRight)) {
    shape.tr = 0
  }

  const cellLeft = peek(map, { row, col }, PeekDirection.Left)
  if (isCellUndefinedOrTechnicallyAlive(cellLeft)) {
    shape.tl = 0
    shape.bl = 0
  }

  const cellRight = peek(map, { row, col }, PeekDirection.Right)
  if (isCellUndefinedOrTechnicallyAlive(cellRight)) {
    shape.tr = 0
    shape.br = 0
  }

  const cellBottomLeft = peek(map, { row, col }, PeekDirection.LowerLeft)
  if (isCellUndefinedOrTechnicallyAlive(cellBottomLeft)) {
    shape.bl = 0
  }

  const cellBelow = peek(map, { row, col }, PeekDirection.Down)
  if (isCellUndefinedOrTechnicallyAlive(cellBelow)) {
    shape.bl = 0
    shape.br = 0
  }

  const cellBottomRight = peek(map, { row, col }, PeekDirection.LowerRight)
  if (isCellUndefinedOrTechnicallyAlive(cellBottomRight)) {
    shape.br = 0
  }

  return shape
}