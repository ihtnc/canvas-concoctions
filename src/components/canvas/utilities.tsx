import {
  RenderLocation,
  type DebugLayerDrawHandler,
  type DebugLayerOptions,
  type DebugLayerRendererValue,
  type Use2DRenderLoopOptions,
  type Coordinates,
  type DebugLayerValue,
} from "./types";

export const DEFAULT_DEBUG_LAYER_OPTIONS: DebugLayerOptions = {
  location: RenderLocation.TopLeft,
  renderFps: true,
  renderSize: true,
  renderClientSize: true,
  renderPixelRatio: true
};

export const DEFAULT_OPTIONS: Use2DRenderLoopOptions = {
  clearEachFrame: true,
  debugLayerRenderer: false
};

export const getDebugLayerRenderer: (value?: DebugLayerRendererValue) => DebugLayerDrawHandler | null = (value) => {
  let options: DebugLayerOptions;
  options = DEFAULT_DEBUG_LAYER_OPTIONS;

  if (value === undefined || value === false) {
    return null;
  }

  if (typeof(value) === "number") {
    options.location = value as RenderLocation;
  }

  const { x, y } = value as Coordinates;
  if (x !== undefined && y !== undefined) {
    options.location = { x, y };
  }

  if (typeof(value) === "object") {
    const {
      location,
      renderFps,
      renderSize,
      renderClientSize,
      renderPixelRatio
    } = value as DebugLayerOptions;

    if (location !== undefined) { options.location = location; }
    if (renderFps !== undefined) { options.renderFps = renderFps; }
    if (renderSize !== undefined) { options.renderSize = renderSize; }
    if (renderClientSize !== undefined) { options.renderClientSize = renderClientSize; }
    if (renderPixelRatio !== undefined) { options.renderPixelRatio = renderPixelRatio; }
  }

  let renderer: DebugLayerDrawHandler;
  if(typeof value === "function") {
    renderer = value;
  }

  const getCoordinates: (value: string, debugValue: DebugLayerValue, context: CanvasRenderingContext2D) => Coordinates = (value, debugValue, context) => {
    let { x, y } = options.location as Coordinates;
    if (x !== undefined && y !== undefined) { return {x, y}; }

    const { width, actualBoundingBoxAscent, actualBoundingBoxDescent } = context.measureText(value);
    const textWidth = width;
    const textHeight = actualBoundingBoxAscent + actualBoundingBoxDescent;
    const offSet = 10;
    const leftX = 0 + offSet;
    const topY = 0 + offSet + offSet;
    const midX = (debugValue.width / 2) - (textWidth / 2);
    const midY = (debugValue.height / 2) - (textHeight / 2);
    const rightX = debugValue.width - textWidth - offSet;
    const bottomY = debugValue.height - textHeight;
    switch(options.location) {
      case RenderLocation.TopLeft:
        x = leftX;
        y = topY;
        break;
      case RenderLocation.TopCenter:
        x = midX;
        y = topY;
        break;
      case RenderLocation.TopRight:
        x = rightX;
        y = topY;
        break;
      case RenderLocation.MiddleLeft:
        x = leftX;
        y = midY;
        break;
      case RenderLocation.Center:
        x = midX;
        y = midY;
        break;
      case RenderLocation.MiddleRight:
        x = rightX;
        y = midY;
        break;
      case RenderLocation.BottomLeft:
        x = leftX;
        y = bottomY;
        break;
      case RenderLocation.BottomCenter:
        x = midX;
        y = bottomY;
        break;
      case RenderLocation.BottomRight:
        x = rightX;
        y = bottomY;
        break;
    }
    return { x, y };
  }

  renderer = (value, context) => {
    const fpsText = options.renderFps ? `fps: ${value.fps}; ` : '';
    const sizeText = options.renderSize ? `size: ${value.width}x${value.height}; ` : '';
    const clientText = options.renderClientSize ? `client: ${value.clientWidth}x${value.clientHeight}; ` : '';
    const ratioText = options.renderPixelRatio ? `ratio: ${value.pixelRatio}; ` : '';
    const debugText = `${fpsText}${sizeText}${clientText}${ratioText}`.trim();
    const { x, y } = getCoordinates(debugText, value, context);
    const originalFillStyle = context.fillStyle;
    context.fillStyle = "#000000";
    context.fillText(debugText, x, y);
    context.fillStyle = originalFillStyle;
  };

  return renderer;
}
