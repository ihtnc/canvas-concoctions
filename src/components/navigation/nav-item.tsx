import { useFloating, useHover, useInteractions } from '@floating-ui/react'
import Image from 'next/image'
import Link from 'next/link'
import { useState } from 'react'

type NavItemProps = {
  href: string,
  className?: string,
  title: string,
  previewHref?: string,
  isActive?: boolean
};

const NavItem = ({ href, className, title, previewHref, isActive }: NavItemProps) => {
  const [isOpen, setIsOpen] = useState(false)
  const { refs, floatingStyles, context } = useFloating({
    open: isOpen,
    onOpenChange: setIsOpen
  })
  const hover = useHover(context)
  const { getReferenceProps, getFloatingProps } = useInteractions([hover])

  return (
    <>
      <Link
        ref={refs.setReference}
        href={href}
        className={`link ${className ?? ''} ${isActive ? 'active': ''}`}
        {...getReferenceProps()}
      >
        {title}
      </Link>
      {!isActive && previewHref && isOpen && (
        <Image
          className='w-72 h-auto border border-black'
          src={previewHref} alt={title}
          width='0' height='0'
          ref={refs.setFloating}
          style={floatingStyles}
          { ...getFloatingProps()}
        />
      )}
    </>
  )
}

export default NavItem