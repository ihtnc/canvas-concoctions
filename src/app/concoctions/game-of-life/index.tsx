'use client';

import { type ConcoctionNavigation } from "@/app/concoctions/utilities";
import {
  type InitRenderHandler,
  type DrawHandler,
  type OnResizeHandler,
  type PreDrawHandler,
  type PostDrawHandler
} from "@/components/canvas/types";
import useAnimatedCanvas from "@/components/canvas/use-animated-canvas";
import { type MatrixCoordinate, type MatrixValue } from "@/utilities/matrix-operations";
import { hexToHSL } from "@/utilities/drawing-operations";
import { type PointerEventHandler, useRef } from "react";
import {
  type CellValue,
  CycleState,
  initialiseCellMap,
  resizeCellMap,
  resetCellMap,
  runCellMapPipeline,
  runCellRenderPipeline,
  getNextCycle,
  isEndPhase
} from "./engine";
import ControlPanel, { type ControlItem } from "@/components/control-panel";
import PlayIcon from "@/components/icons/play-icon";
import PauseIcon from "@/components/icons/pause-icon";
import ForwardIcon from "@/components/icons/forward-icon";
import TrashIcon from "@/components/icons/trash-icon";

type GameOfLifeProps = {
  className?: string,
  cellSize?: number,
  autoStart?: boolean,
  aliveColor?: string,
  dyingColor?: string,
  growingColor?: string
};
type DefaultData = {
  MinCellSize: number,
  MaxCellSize: number,
  DefaultCellSize: number,
  DefaultAutoStart: boolean,
  DefaultAliveColor: string,
  DefaultDyingColor: string,
  DefaultGrowingColor: string
}
enum ButtonType {
  Start,
  Pause,
  Step
}

const DEFAULT_DATA: DefaultData = {
  MinCellSize: 10,
  MaxCellSize: 30,
  DefaultCellSize: 10,
  DefaultAutoStart: true,
  DefaultAliveColor: '#09AB00',
  DefaultDyingColor: '#C0C0C0',
  DefaultGrowingColor: '#0BD100',
}

const GameOfLife = ({
  className,
  cellSize=DEFAULT_DATA.DefaultCellSize,
  autoStart=DEFAULT_DATA.DefaultAutoStart,
  aliveColor=DEFAULT_DATA.DefaultAliveColor,
  dyingColor=DEFAULT_DATA.DefaultDyingColor,
  growingColor=DEFAULT_DATA.DefaultGrowingColor
}: GameOfLifeProps) => {
  const cellMap = useRef<MatrixValue<CellValue> | null>(null);
  const newCellCoordinate: MatrixCoordinate = { row: 0, col: 0 };

  let canGenerateCell: boolean = false;
  let newCellInitiated: boolean = false;

  let cycleState: CycleState = autoStart ? CycleState.Start : CycleState.Stop;
  let lastButtonClicked: ButtonType = autoStart ? ButtonType.Start : ButtonType.Pause;

  if (cellSize < DEFAULT_DATA.MinCellSize) { cellSize = DEFAULT_DATA.MinCellSize; }
  else if (cellSize > DEFAULT_DATA.MaxCellSize) { cellSize = DEFAULT_DATA.MaxCellSize; }

  const aliveHSLProp = hexToHSL(aliveColor);
  const aliveHSL = aliveHSLProp === undefined
    ? hexToHSL(DEFAULT_DATA.DefaultAliveColor)!
    : aliveHSLProp!;

  const dyingHSLProp = hexToHSL(dyingColor);
  const dyingHSL = dyingHSLProp === undefined
    ? hexToHSL(DEFAULT_DATA.DefaultDyingColor)!
    : dyingHSLProp!;

  const growingHSLProp = hexToHSL(growingColor);
  const growingHSL = growingHSLProp === undefined
    ? hexToHSL(DEFAULT_DATA.DefaultGrowingColor)!
    : growingHSLProp!;

  let cycleIndex: number = 0;

  const initFn: InitRenderHandler = (canvas) => {
    const row = Math.floor(canvas.height / cellSize);
    const col = Math.floor(canvas.width / cellSize);
    const map = initialiseCellMap(row, col);
    cellMap.current = map;
  };

  const onResizeFn: OnResizeHandler = (canvas, width, height) => {
    if (cellMap?.current === null) { return; }

    const newRow = Math.floor(height / cellSize);
    const newCol = Math.floor(width / cellSize);

    const newSize = resizeCellMap(cellMap.current, newRow, newCol);
    cellMap.current = newSize;
  };

  const startNewCells: PointerEventHandler<HTMLCanvasElement> = (event) => {
    canGenerateCell = true;
    updateNewCellCoordinate(event);

    if (lastButtonClicked === ButtonType.Start) {
      cycleState = CycleState.Stop;
    }

    newCellInitiated = true;
  }

  const stopNewCells: PointerEventHandler<HTMLCanvasElement> = (event) => {
    canGenerateCell = false;

    if (newCellInitiated && lastButtonClicked === ButtonType.Start) {
      cycleState = CycleState.Start;
    }

    newCellInitiated = false;
  }

  const updateNewCellCoordinate: PointerEventHandler<HTMLCanvasElement> = (event) => {
    const canvas = event.currentTarget;
    const rect = canvas.getBoundingClientRect();
    const x = event.clientX - rect.left;
    const y = event.clientY - rect.top;
    const col = Math.floor(x / cellSize);
    const row = Math.floor(y / cellSize);

    newCellCoordinate.row = row;
    newCellCoordinate.col = col;
  }

  const predrawFn: PreDrawHandler = (canvas, data) => {
    if (cellMap.current === null) { return; }

    const map = cellMap.current;
    const newMap = runCellMapPipeline(map, cycleIndex, canGenerateCell ? newCellCoordinate : undefined);
    cellMap.current = newMap;
  };

  const drawFn: DrawHandler = ({ context }) => {
    if (cellMap?.current === null) { return; }

    runCellRenderPipeline(
      context,
      {
        map: cellMap.current,
        width: cellSize, height: cellSize,
        aliveColor: aliveHSL, dyingColor: dyingHSL, growingColor: growingHSL,
        cycleIndex
      },
      [],
      []
    );
  };

  const postDrawFn: PostDrawHandler = () => {
    if (lastButtonClicked === ButtonType.Step && isEndPhase(cycleIndex)) {
      cycleState = CycleState.Stop;
    }
    cycleIndex = getNextCycle(cycleIndex, cycleState);
  };

  const startCycle = () => {
    lastButtonClicked = ButtonType.Start;
    cycleState = CycleState.Start;
  }

  const pauseCycle = () => {
    lastButtonClicked = ButtonType.Pause;
    cycleState = CycleState.Stop;
  }

  const stepCycle = () => {
    lastButtonClicked = ButtonType.Step;
    cycleState = CycleState.Start;
  }

  const resetConcoction = () => {
    if (cellMap?.current === null) { return; }

    canGenerateCell = false;
    resetCellMap(cellMap.current);
  }

  const { Canvas } = useAnimatedCanvas({
    init: initFn,
    predraw: predrawFn,
    draw: drawFn,
    postdraw: postDrawFn,
    onResize: onResizeFn
  });

  const hidePlayIcon = (): boolean => {
    return lastButtonClicked === ButtonType.Start;
  };

  const hidePauseIcon = (): boolean => {
    const showNormal = lastButtonClicked === ButtonType.Start;
    return !showNormal;
  }

  const controls: Array<ControlItem> = [{
    type: "button",
    onClickHandler: startCycle,
    content: (<PlayIcon />),
    title: "Start cycle",
    hidden: hidePlayIcon
  }, {
    type: "button",
    onClickHandler: pauseCycle,
    content: (<PauseIcon />),
    title: "Pause cycle",
    hidden: hidePauseIcon
  }, {
    type: "button",
    onClickHandler: stepCycle,
    content: (<ForwardIcon />),
    title: "Forward 1 cycle"
  }, {
    type: "button",
    onClickHandler: resetConcoction,
    content: (<TrashIcon />),
    title: "Reset canvas",
    className: "ml-auto"
  }];

  return <div className="flex flex-col w-full h-full gap-2">
    <Canvas
      className={className}
      onPointerDown={startNewCells}
      onPointerUp={stopNewCells}
      onPointerOut={stopNewCells}
      onPointerMove={updateNewCellCoordinate}
    />
    <ControlPanel controls={controls} className="w-full" />
  </div>
};

export const NavigationDetails: ConcoctionNavigation = {
  linkTitle: 'Game of Life',
  linkUrl: 'game-of-life',
  title: 'Game of Life',
  previewUrl: '/previews/game-of-life.gif'
};

export default GameOfLife;