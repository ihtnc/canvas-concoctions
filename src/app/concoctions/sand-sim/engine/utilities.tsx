import { type ParticleColorMap, type ParticleValue } from "./types"
import ENGINE_DATA from './data'
import { type MatrixCoordinate, type MatrixValue, PeekDirection, peek } from "@/utilities/matrix-operations"
import { chooseRandom, deepCopy } from "@/utilities/misc-operations"
import { areHSLsEqual } from "@/utilities/drawing-operations"

export const isValueUndefinedOrNotEmpty = (particle?: ParticleValue): boolean => {
  return particle === undefined || !isValueEmpty(particle)
}

export const isValueEmpty = (particle: ParticleValue): boolean => {
  return particle.color === undefined || particle.value === ENGINE_DATA.EmptyParticle.value
}

export const calculateDistance = (time: number): number => {
  // distance = (initialVelocity * time) + (0.5 * acceleration * time^2)
  const initial = time * ENGINE_DATA.InitialVelocity
  const accelaration = (0.5 * ENGINE_DATA.Acceleration * time * time)
  return initial + accelaration
}

export const getNextCoordinate = (value: MatrixValue<ParticleValue>, current: MatrixCoordinate): MatrixCoordinate | undefined => {
  let direction: PeekDirection = PeekDirection.Down
  let retryCount: number = 0

  while (retryCount < 2) {
    const particleBelow = peek(value, current, direction)
    if (particleBelow === undefined) { return undefined }

    if (isValueEmpty(particleBelow)) {
      const next: MatrixCoordinate = {
        row: current.row + 1,
        col: current.col
      }

      if (direction === PeekDirection.LowerLeft) { next.col = current.col - 1 }
      else if (direction === PeekDirection.LowerRight) { next.col = current.col + 1 }

      return next
    }

    const option = chooseRandom(1, 100)
    if (option <= 48) { direction = PeekDirection.LowerLeft }
    else if (option <= 52) { return undefined }
    else { direction = PeekDirection.LowerRight }

    retryCount++
  }

  return undefined
}

export const getNextValue = (particle: ParticleValue): ParticleValue => {
  const nextValue = deepCopy(particle)
  nextValue.value = nextValue.value + ENGINE_DATA.IncrementValue
  return deepCopy(nextValue)
}

export const getColorMap = (value: MatrixValue<ParticleValue>): Array<ParticleColorMap> => {
  const sorted: Array<ParticleColorMap> = []

  for (let i = 0; i < value.length; i++) {
    for (let j = 0; j < value[i].length; j++) {
      if (value[i][j].color === undefined) { continue }

      const color = value[i][j].color!
      let group = sorted.find(item => areHSLsEqual(item.color, color))
      if (group === undefined) {
        const newLength = sorted.push({ color: deepCopy(color), particles: [] })
        group = sorted[newLength - 1]
      }
      group.particles.push({ row: i, col: j })
    }
  }

  return sorted
}