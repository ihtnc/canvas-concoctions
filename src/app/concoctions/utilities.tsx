import { NavigationDetails as sandSimNav } from './sand-sim';
import { NavigationDetails as tagVisualiserNav } from './tag-visualiser';
import { NavigationDetails as gameOfLifeNav } from './game-of-life';

export type ConcoctionNavigation = {
  linkTitle: string,
  linkUrl: string,
  title: string,
  previewUrl?: string
};

type GetConcoctionsResponse = Array<ConcoctionNavigation>;
export const getConcoctions = (): GetConcoctionsResponse => {
  return [
    sandSimNav,
    tagVisualiserNav,
    gameOfLifeNav
  ];
};
