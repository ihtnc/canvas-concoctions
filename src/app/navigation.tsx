'use client';

import { getConcoctions } from './concoctions/utilities';
import Link from 'next/link';
import { usePathname } from 'next/navigation';

const Navigation = () => {
  const links = getConcoctions();
  const pathName = usePathname();

  const activeLink = links.find(l => `/concoctions/${l.linkUrl}` === pathName);
  const isHomeActive = pathName === '/';

  return (
    <nav>
      <ul className='flex flex-row gap-2'>
        <li key='/' className='flex'>
          <Link
            href='/'
            className={`link ${isHomeActive ? 'active' : ''}`}>
            Home
          </Link>
        </li>
        {links.map((value) => (
          <li key={value.linkUrl} className='flex'>
            <Link
              href={`/concoctions/${value.linkUrl}`}
              className={`link ${value.linkUrl === activeLink?.linkUrl ? 'active' : ''}`}>
              {value.linkTitle}
            </Link>
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default Navigation;