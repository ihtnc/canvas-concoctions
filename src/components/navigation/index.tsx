'use client'

import { usePathname, useSearchParams } from 'next/navigation'
import { getConcoctions } from '@/app/concoctions/utilities'
import NavItem from './nav-item'

type NavigationProps = {
  baseUrl?: string
};

const Navigation = ({ baseUrl }: NavigationProps) => {
  const links = getConcoctions()
  const pathName = usePathname()
  const concoctionId = pathName.split('/').pop()
  const searchParams = useSearchParams()
  const isApp = searchParams.has("app")
  const itemBaseUrl = baseUrl !== undefined ? `/${baseUrl.replace(/^\//, '').replace(/\/$/, '')}` : ''

  const constructPath = (linkUrl: string) => {
    const trimUrl = linkUrl.replace(/^\//, '')
    return `${itemBaseUrl}/${trimUrl}`
  }

  return isApp === false ? (
    <section className="flex flex-col self-center">
      <nav>
        <ul className='flex flex-row gap-2'>
          <li key='/' className='flex'>
            <NavItem
              href='/'
              title='Home'
              isActive={pathName === '/'} />
          </li>
          {links.map((value) => (
            <li key={value.linkUrl} className='flex'>
              <NavItem
                href={`${constructPath(value.linkUrl)}`}
                title={value.linkTitle}
                previewHref={value.previewUrl ? value.previewUrl : '/previews/default.gif'}
                isActive={value.linkUrl === concoctionId}
              />
            </li>
          ))}
        </ul>
      </nav>
    </section>
  ) : null
}

export default Navigation