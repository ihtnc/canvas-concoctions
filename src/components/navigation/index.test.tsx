import { render, screen } from '@testing-library/react'
import Navigation from './index'
import { Mock, afterEach, beforeEach, describe, expect, test, vi } from 'vitest'
import { usePathname } from 'next/navigation'
import { getConcoctions } from '@/app/concoctions/utilities'

vi.mock('next/navigation', () => ({
  usePathname: vi.fn()
}))

vi.mock('@/app/concoctions/utilities', () => ({
  getConcoctions: vi.fn()
}))

vi.mock('./nav-item', () => ({
  default: (props: any) => <div>{JSON.stringify(props)}</div>
}))

describe('Navigation component', () => {
  let usePathnameMock: Mock
  let getConcoctionsMock: Mock

  beforeEach(() => {
    usePathnameMock = usePathname as Mock
    usePathnameMock.mockReturnValue('')

    getConcoctionsMock = getConcoctions as Mock
    getConcoctionsMock.mockReturnValue([])
  })

  afterEach(() => {
    vi.resetAllMocks()
  })

  test('should retrieve list of concoctions', () => {
    render(<Navigation />)

    expect(getConcoctionsMock).toHaveBeenCalled()
  })

  test('should retrieve current path', () => {
    render(<Navigation />)

    expect(usePathnameMock).toHaveBeenCalled()
  })

  test('should render home menu and list of concoctions', () => {
    getConcoctionsMock.mockReturnValue([
      { linkUrl: 'concoction1', linkTitle: 'Concoction 1' },
      { linkUrl: 'concoction2', linkTitle: 'Concoction 2', previewUrl: 'concoction2.jpg' }
    ])

    const { container } = render(<Navigation />)

    expect(container).toMatchSnapshot()
  })

  test('should set home menu item as active when current path is on root', () => {
    usePathnameMock.mockReturnValue('/')

    const { container } = render(<Navigation />)

    expect(container).toMatchSnapshot()
  })

  test('should set appropriate concoction menu item as active based on current path', () => {
    usePathnameMock.mockReturnValue('/concoction2')

    getConcoctionsMock.mockReturnValue([
      { linkUrl: 'concoction1', linkTitle: 'Concoction 1' },
      { linkUrl: 'concoction2', linkTitle: 'Concoction 2' }
    ])

    const { container } = render(<Navigation />)

    expect(container).toMatchSnapshot()
  })

  test('should render menu items if current path is not recognised', () => {
    usePathnameMock.mockReturnValue('/concotion3')

    getConcoctionsMock.mockReturnValue([
      { linkUrl: 'concoction1', linkTitle: 'Concoction 1' },
      { linkUrl: 'concoction2', linkTitle: 'Concoction 2' }
    ])

    const { container } = render(<Navigation />)

    expect(container).toMatchSnapshot()
  })

  test('should append baseUrl to concoction linkUrl', () => {
    usePathnameMock.mockReturnValue('/concoction1')

    getConcoctionsMock.mockReturnValue([
      { linkUrl: 'concoction1', linkTitle: 'Concoction 1' }
    ])

    const { container } = render(<Navigation baseUrl='/base' />)

    expect(container).toMatchSnapshot()
  })

  test.each([
    { baseUrl: 'base', linkUrl: 'concoction', expected: '/base/concoction' },
    { baseUrl: '/base', linkUrl: '/concoction', expected: '/base/concoction' },
    { baseUrl: '/base', linkUrl: 'concoction', expected: '/base/concoction' },
    { baseUrl: 'base', linkUrl: '/concoction', expected: '/base/concoction' },
  ])('should handle URLs without leading / (baseUrl: $baseUrl; linkUrl: $linkUrl)', ({ baseUrl, linkUrl, expected }: { baseUrl: string, linkUrl: string, expected: string }) => {
    usePathnameMock.mockReturnValue('/')

    const concoctionUrl = 'concoction'
    getConcoctionsMock.mockReturnValue([
      { linkUrl, linkTitle: 'Concoction 1' }
    ])

    render(<Navigation baseUrl={baseUrl} />)

    const content = screen.getByText(expected, { exact: false })

    expect(content).toBeDefined()
  })
})