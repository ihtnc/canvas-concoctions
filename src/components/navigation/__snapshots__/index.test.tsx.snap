// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navigation component > should append baseUrl to concoction linkUrl 1`] = `
<div>
  <section
    class="flex flex-col self-center"
  >
    <nav>
      <ul
        class="flex flex-row gap-2"
      >
        <li
          class="flex"
        >
          <div>
            {"href":"/","title":"Home","isActive":false}
          </div>
        </li>
        <li
          class="flex"
        >
          <div>
            {"href":"/base/concoction1","title":"Concoction 1","previewHref":"/previews/default.gif","isActive":true}
          </div>
        </li>
      </ul>
    </nav>
  </section>
</div>
`;

exports[`Navigation component > should not render anything if app search params exists 1`] = `<div />`;

exports[`Navigation component > should render home menu and list of concoctions 1`] = `
<div>
  <section
    class="flex flex-col self-center"
  >
    <nav>
      <ul
        class="flex flex-row gap-2"
      >
        <li
          class="flex"
        >
          <div>
            {"href":"/","title":"Home","isActive":false}
          </div>
        </li>
        <li
          class="flex"
        >
          <div>
            {"href":"/concoction1","title":"Concoction 1","previewHref":"/previews/default.gif","isActive":false}
          </div>
        </li>
        <li
          class="flex"
        >
          <div>
            {"href":"/concoction2","title":"Concoction 2","previewHref":"concoction2.jpg","isActive":false}
          </div>
        </li>
      </ul>
    </nav>
  </section>
</div>
`;

exports[`Navigation component > should render menu items if current path is not recognised 1`] = `
<div>
  <section
    class="flex flex-col self-center"
  >
    <nav>
      <ul
        class="flex flex-row gap-2"
      >
        <li
          class="flex"
        >
          <div>
            {"href":"/","title":"Home","isActive":false}
          </div>
        </li>
        <li
          class="flex"
        >
          <div>
            {"href":"/concoction1","title":"Concoction 1","previewHref":"/previews/default.gif","isActive":false}
          </div>
        </li>
        <li
          class="flex"
        >
          <div>
            {"href":"/concoction2","title":"Concoction 2","previewHref":"/previews/default.gif","isActive":false}
          </div>
        </li>
      </ul>
    </nav>
  </section>
</div>
`;

exports[`Navigation component > should set appropriate concoction menu item as active based on current path 1`] = `
<div>
  <section
    class="flex flex-col self-center"
  >
    <nav>
      <ul
        class="flex flex-row gap-2"
      >
        <li
          class="flex"
        >
          <div>
            {"href":"/","title":"Home","isActive":false}
          </div>
        </li>
        <li
          class="flex"
        >
          <div>
            {"href":"/concoction1","title":"Concoction 1","previewHref":"/previews/default.gif","isActive":false}
          </div>
        </li>
        <li
          class="flex"
        >
          <div>
            {"href":"/concoction2","title":"Concoction 2","previewHref":"/previews/default.gif","isActive":true}
          </div>
        </li>
      </ul>
    </nav>
  </section>
</div>
`;

exports[`Navigation component > should set home menu item as active when current path is on root 1`] = `
<div>
  <section
    class="flex flex-col self-center"
  >
    <nav>
      <ul
        class="flex flex-row gap-2"
      >
        <li
          class="flex"
        >
          <div>
            {"href":"/","title":"Home","isActive":true}
          </div>
        </li>
      </ul>
    </nav>
  </section>
</div>
`;
